zenoh/src/api/publisher.rs

```diff
@@ -350,6 +358,13 @@ impl<'a> Publisher<'a> {
     pub fn undeclare(self) -> impl Resolve<ZResult<()>> + 'a {
         Undeclarable::undeclare_inner(self, ())
     }
+
+    fn undeclare_matching_listeners(&self) -> ZResult<()> {
+        for id in zlock!(self.matching_listeners).drain() {
```

Comment:
This code would be safer w.r.t. mutex deadlock if:

1. it takes the lock
2. it drains the iterator in a supporting vec
3. it releases the lock
4. it undeclares the matching listeners

Reply:
#1116